api/service/model_endpoint_service_test.go

```diff
@@ -377,6 +419,44 @@ func Test_modelEndpointsService_UpdateEndpoint(t *testing.T) {
 			nil,
 			true,
 		},
+		{
+			name: "failure: save to database",
+			fields: fields{
+				istioClients:           map[string]istio.Client{env.Name: &istioCliMock.Client{}},
+				modelEndpointStorage:   &storageMock.ModelEndpointStorage{},
+				versionEndpointStorage: &storageMock.VersionEndpointStorage{},
+				environment:            testEnvironmentName,
+				observabilityEventProducer: func() event.EventProducer {
+					eProducer := &eventMock.EventProducer{}
+					eProducer.On("ModelEndpointChangeEvent", mock.Anything, mock.Anything).Return(nil)
+					return eProducer
+				}(),
+			},
+			mockFunc: func(s *modelEndpointsService) {
+				vs, _ := s.createVirtualService(model1, updatedModelEndpointReq)
+
+				mockIstio := s.istioClients[env.Name].(*istioCliMock.Client)
+				mockIstio.On("GetVirtualService", context.Background(), "project-1", "model-1").Return(vs, nil)
+
+				mockIstio.On("PatchVirtualService", context.Background(), "project-1", vs).Return(vs, nil).Once()
+
+				mockVeStorage := s.versionEndpointStorage.(*storageMock.VersionEndpointStorage)
+				mockVeStorage.On("Get", updatedModelEndpointReq.Rule.Destination[0].VersionEndpointID).Return(newVersionEndpoint, nil)
+
+				mockMeStorage := s.modelEndpointStorage.(*storageMock.ModelEndpointStorage)
+				mockMeStorage.On("Save", context.Background(), mock.AnythingOfType("*models.ModelEndpoint"), mock.AnythingOfType("*models.ModelEndpoint")).Return(fmt.Errorf("failed to save"))
+
+				mockIstio.On("PatchVirtualService", context.WithoutCancel(context.Background()), "project-1", vs).Return(vs, nil).Once()
```

Comment:
ultra nit: lol this test doesn't really check if the `vs` object used for the rollback is the the original `vs` that was retrieved from `GetVirtualService` before it gets patched. In other words, we currently have

1. `GetVirtualService` returns `vs` 
2. `PatchVirtualService` patches `vs` with `vs` (ideally the new `vs` object should be somewhat different)
3. Roll back with another `PatchVirtualService` with the same `vs` again (ideally this should be the `vs` from step 1 and not the same `vs` in step 2)

Functionally the test checks all the functions that need to be called in such a scenario though, so everything else is good 😅